pkg/porter/publish.go

```diff
@@ -0,0 +1,172 @@
+package porter
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/deislabs/porter/pkg/config"
+	"github.com/docker/cli/cli/command"
+	cliflags "github.com/docker/cli/cli/flags"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/pkg/jsonmessage"
+	"github.com/docker/docker/pkg/term"
+	"github.com/docker/docker/registry"
+	"github.com/pkg/errors"
+)
+
+// PublishOptions are options that may be specified when publishing a bundle.
+// Porter handles defaulting any missing values.
+type PublishOptions struct {
+	File     string
+	Tag      string
+	CNABFile string
+}
+
+func (p PublishOptions) Validate(porter *Porter) error {
+	if p.File != "" {
+		return nil
+	}
+	fs := porter.Context.FileSystem
+	pwd, err := os.Getwd()
+	if err != nil {
+		return errors.Wrap(err, "could not get current working directory")
+	}
+	fmt.Println(pwd)
+	files, err := fs.ReadDir(pwd)
+	if err != nil {
+		return errors.Wrapf(err, "could not list current directory %s", pwd)
+	}
+	foundManifest := false
+	for _, f := range files {
+		if !f.IsDir() && f.Name() == config.Name {
+			foundManifest = true
+		}
+	}
+
+	if p.File == "" && !foundManifest {
+		return errors.New("could not find porter.yaml. run `porter create` and `porter build` to create a new bundle before publishing")
+	}
+
+	if p.Tag != "" {
+		_, err := reference.ParseNormalizedNamed(p.Tag)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Publish is a composite function that publishes an invocation image, rewrites the porter manifest
+// and then regenerates the bundle.json. Finally it [TODO] publishes the manifest to an OCI registry.
+func (p *Porter) Publish(opts PublishOptions) error {
+	var err error
+	if opts.File != "" {
+		err = p.Config.LoadManifestFrom(opts.File)
+	} else {
+		err = p.Config.LoadManifest()
+	}
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	cli, err := p.getDockerClient(ctx)
+	if err != nil {
+		return err
+	}
+	if opts.Tag != "" {
```

Comment:
Yeah this entire part I think should be removed since tag = bundle tag, not invocation image tag. Feel free to give things better names anywhere you like. 😀

Reply:
I think we should still validate that the tag given is a valid OCI tag no?

Reply:
Yes, in the `PublishOptions.Validate` method. I was just referring to the lines 81-83 here where we are changing the invocation image tag.